Feature/add pextra

joakimkjellsson opened the pull request:
Hi @seb-wahl 

I've added functionality to activate PEXTRA in OpenIFS, which means online tendencies such as "u-tendency due to convection", "T tendency due to diffusion" etc. 
It took a few careful changes, but seems to work now. 

I'd to merge this into the geomar-setup branch, but I'd like your blessing first ;-) 

Cheers
Joakim

Commit:
add my test script for blogin

Commit:
Merge branch 'feature/geomar-setup' of https://www.github.com/esm-tools/esm_tools into feature/geomar-setup

Commit:
running with pextra crashes at end

Commit:
pextra now works on blogin

seb-wahl commented on configs/components/oifs/oifs.yaml:
```diff
@@ -1044,10 +1059,11 @@ preprocess:
-#                type: shell
+postprocess:
+        postprocess_shell:
+                method: "${general.esm_namelist_dir}/../configs/components/oifs/oifs-43r3-postprocess.sh ${work_dir} ECE3 ${start_date!syear!smonth!sday} ${end_date!syear!smonth!sday}"
```
This only works if you have a .esmtoolsrc file (which we actually abandoned with v5). That is why it is commented. I talk to the AWI people asap to see whether they have a solution. I've tried to solve it in the past, but it hasn't worked.

denizural replied:
Yes, we have a solution. After `.esmtoolsrc` becoming obsolete, I added some get/set functionality to the `esm_versions` command. Let's have a look at these examples:

$ esm_versions get esm_tools
```
esm_tools
---------
Version:         5.0.5
File Path:       /mnt/lustre01/pf/a/a271096/esm_all_packages/esm_tools
Branch:          hotfix/fix_lang_issues
Git Describe:    v5.0.8-8-g4ce568f
```

$ esm_versions get esm_tools **file_path**
`/mnt/lustre01/pf/a/a271096/esm_all_packages/esm_tools`

Let's use **command substitution** in shell
``ls `esm_versions get esm_tools file_path`/namelists/oifs``
or 
`ls $(esm_versions get esm_tools file_path)/namelists/oifs`

This yields
`40r1  43r3  namelist.oifs.esm.sh`

and 

`$ file $(esm_versions get esm_tools file_path)/configs/components/oifs/oifs-43r3-postprocess.sh`
gives 
`/some/long/path/esm_tools/configs/components/oifs/oifs-43r3-postprocess.sh: Bourne-Again shell script text executable`

I think this is what you are looking for. @seb-wahl: we can discuss that this morning at the meeting. Have a nice week!

joakimkjellsson replied:
Hi @denizural @seb-wahl 

"general.esm_namelist_dir" occurs three times in oifs.yaml, and I replaced all of them with "$(esm_versions get esm_tools file_path)/namelists/". 

```bash
namelist_dir: "$(esm_versions get esm_tools file_path)/namelists/oifs/${version}/pictl/"
```

The preprocessing seems ok, but the namelist fort.4 can't be found. For example: 

```bash
(base) blogin1:~/esm_tools/runscripts/oifs $ esm_runscripts oifs-43r3-tco95-amip-blogin-test_JK.yaml -e oifs-test5-23 -c 
First year, checking if we need to compile...

================================================================================
STARTING SIMULATION JOB!
Experiment ID = oifs-test5-23
Setup = oifs
Experiment is installed in:
       /scratch/usr/shkjocke/esm-slask//oifs-test5-23
================================================================================

File not found: $(esm_versions get esm_tools file_path)/namelists/oifs/43r3/pictl/fort.4...
```

Seems that ```$(esm_versions get esm_tools file_path)``` is not expanded as expected. Can we solve this somehow?

Cheers
Joakim

seb-wahl replied:
Sorry I think I put you on the wrong path. @denizural @joakimkjellsson I have a look at this now... Postprocessing does not work because there is no postprocessing plugin, I think. I let you know once I have a solution

Commit:
swapped esm_namelist_dir for command that does not work

seb-wahl committed:
added postprocessing (not yet working)
made oifs.yaml really independent from .esmtoolsrc by setting path
for preprocessing as follows
preprocess:
        preprocess_shell:
		                  method: "$(esm_versions get esm_tools file_path)/configs....

seb-wahl commented:
Ok I solved the variables issue, but NOT the postprocessing. ESM-Tools can still resolve the `esm_namelist_dir` variable but NOT in the plugins, because they are used BEFORE those variables are set (I think). 
I need to talk to @denizural, @mandresm or @dbarbi to figure out how we can correctly trigger postprocessing, see also my question on slack.
To get back what you had @joakimkjellsson, pull my latest updates on feature/add-pextra. It also makes the preprocessing really independent of the `.esmtoolsrc`. If you have questions, we can talk tomorrow morning any time.

joakimkjellsson commented:
Hi @seb-wahl 
I can't find your updates to the add-pextra branch. Did you push it already? 
Here's what happens for me on blogin: 
```bash
(base) blogin1:~/esm_tools/runscripts/oifs $ git pull
warning: redirecting to https://github.com/esm-tools/esm_tools.git/
Already up to date.
```

seb-wahl commented:
`git push` always helps. Done. Had a chat with @denizural He works on it already.

joakimkjellsson commented:
@seb-wahl 
I got your changes, and it works as long as I keep my .esmtoolsrc
If I remove it, then I'm back to the old

```bash
(base) blogin1:~/esm_tools/runscripts/oifs $ esm_runscripts oifs-43r3-tco95-amip-blogin-test_JK.yaml  -e oifs-test5-25 -c
/home/shkjocke/esm_tools/runscripts/oifs/oifs-43r3-tco95-amip-blogin-test_JK.yaml
First year, checking if we need to compile...

================================================================================
STARTING SIMULATION JOB!
Experiment ID = oifs-test5-25
Setup = oifs
Experiment is installed in:
       /scratch/usr/shkjocke/esm-slask//oifs-test5-25
================================================================================

File not found: NONE_YET/oifs/43r3/fort.4...
```

I'm fine with this, but it does mean this version still relies on .esmtoolsrc to exist. 

Cheers
Joakim

seb-wahl commented:
This shouldn't be the case. I have a look and let you know. Had a chat with Paul this morning. Including your postprocessing shell script requires a bit more work on my side as I need to understand the deep internal of `esm_runscripts` to hook in your script correctly.

Commit:
corrected oifs namelist and runscript to continue old Tco199 AMIP run with ESM-Tools release 5

joakimkjellsson commented:
Hi @seb-wahl 
I've set post_processing: 0 in my runscript now and it seems ok. 
This pull request is about making PEXTRA run and that seems to work. 
Should we just merge this stuff into our geomar-setup branch. Then we can open a new branch to work on postprocessing etc? 

Cheers
Joakim

seb-wahl commented:
I'll merge it now as the branch was about pextra and not solving all other problems :-)
